### PR TITLE
feat: add pg for db adapter

### PR DESCRIPTION
**Pull request by Kininaru**
Resolved: https://github.com/casbin/casdoor/issues/45

Deployed at: https://door.chromium.link/

This demo uses postgresql.

### CODE REVIEW

File: conf/app.conf [hunk 1]
```diff
@@ -3,6 +3,7 @@ httpport = 8000
 runmode = dev
 SessionOn = true
 copyrequestbody = true
+db = "mysql"
```
Comment (hsluoyz) [inline conf/app.conf:6]: no quotes

File: object/adapter.go [hunk 2]
```diff
@@ -19,7 +19,8 @@ import (
 	"runtime"
 
 	"github.com/astaxie/beego"
-	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/go-sql-driver/mysql" // mysql
+	_ "github.com/lib/pq"              // postgres
```
Comment (hsluoyz) [inline object/adapter.go:23]: comment pg

File: go.mod [hunk 3]
```diff
@@ -13,8 +13,8 @@ require (
 	github.com/go-sql-driver/mysql v1.5.0
 	github.com/google/uuid v1.2.0
 	github.com/jinzhu/configor v1.2.1 // indirect
+	github.com/lib/pq v1.8.0
```
Comment (hsluoyz) [inline go.mod:16]: Don't add pg dependency

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: conf/app.conf (hunk 1 above)
Comment (Kininaru) [reply conf/app.conf:6]: Used here: https://github.com/casbin/casdoor/pull/50/commits/4dc5b21c02b6ff7f98657f5cbfc9f6bd510e9b15#diff-009ed459acbafa374d33f30ee04034a1bf95db713f5abf5a0a45cfe9545c343fR30 
and here: https://github.com/casbin/casdoor/pull/50/commits/4dc5b21c02b6ff7f98657f5cbfc9f6bd510e9b15#diff-6e87f35c571ba4718e96ba663dec842335060094fe60f9066b2aaa04570f2c1dR39

File: object/adapter.go (hunk 2 above)
Comment (Kininaru) [reply object/adapter.go:23]: Users should fill `app.conf` with `db = "postgres"`, or xorm can't connect to pg. So I suggest to use `postgres` here.

File: go.mod (hunk 3 above)
Comment (Kininaru) [reply go.mod]: Removed that.

File: conf/app.conf (hunk 1 above)
Comment (hsluoyz) [reply conf/app.conf:6]: `db = "mysql"` -> `db = mysql`

Comment (hsluoyz) [reply conf/app.conf:6]: " is quote

**Comment by hsluoyz** [comment]
@kuchaguangjie plz review.

File: object/adapter.go [hunk 4]
```diff
@@ -19,7 +19,8 @@ import (
 	"runtime"
 
 	"github.com/astaxie/beego"
-	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/go-sql-driver/mysql" // db = mysql
+	// _ "github.com/lib/pq"              db = postgres
```
Comment (hsluoyz) [inline object/adapter.go:23]: // db = postgres

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/adapter.go (hunk 4 above)
Comment (hsluoyz) [reply object/adapter.go]: @Kininaru didn't fix

**Comment by kuchaguangjie** [comment]
After pulling newest code on master branch, and change `conf/app.conf`, set `dataSourceName` to:

>  dataSourceName = "user=postgres password=xxx host=127.0.0.1 port=5432 dbname=casdoor sslmode=disable"

Then start, would get error:

> panic: invalid DSN: missing the slash separating the database name
> 
> goroutine 1 [running]:
> github.com/casdoor/casdoor/object.(*Adapter).open(0xc0004596b0)
> 	/mnt/star/git_repository/casdoor/object/adapter.go:86 +0x1a5
> github.com/casdoor/casdoor/object.NewAdapter(0xc00013cd36, 0x5, 0xc00010e180, 0x55, 0x55)
> 	/mnt/star/git_repository/casdoor/object/adapter.go:64 +0x7d
> github.com/casdoor/casdoor/object.InitAdapter()
> 	/mnt/star/git_repository/casdoor/object/adapter.go:39 +0xab
> main.main()
> 	/mnt/star/git_repository/casdoor/main.go:29 +0x26
> exit status 2

So, what is the expected `dataSourceName` for pg ?

**Comment by Kininaru** [comment]
@kuchaguangjie 

In app.conf, set: `db = postgres`,  `dataSourceName = "user=postgres password=xxx host=127.0.0.1 port=5432 sslmode=disable dbname="`, and set dbName to `casdoor`.

In object/adapter.go: enable line 23

Since postgres has some differences from other DBs in xorm, we can't automatically create a database in Go code. This means that you should create a database for Casdoor manually before starting it.

**Comment by hsluoyz** [comment]
`dataSourceName` is Xorm's driverName, see Xorm docs for possible values: https://gobook.io/read/gitea.com/xorm/manual-en-US/chapter-01/2.engine_group.html

You can use `postgres` for it.

@Kininaru we need to document this in README.

**Comment by Kininaru** [comment]
@hsluoyz Got that. I'll make a pull request about this later today.

@kuchaguangjie This is my app.conf. I'm using postgres now, and it works well. Hope it'll help you.

```
appname = casdoor
httpport = 8000
runmode = dev
SessionOn = true
copyrequestbody = true
db = postgres
dataSourceName = "user=postgres password=xxx sslmode=disable dbname="
dbName = casdoor
AuthState = "casdoor"
UseProxy = false
EnableDocs = true
```

**Comment by kuchaguangjie** [comment]
@Kininaru  According to your configuration, it works for pg now, thank you.

**Comment by hsluoyz** [comment]
@Kininaru plz put this into README: https://github.com/casbin/casdoor#config

Provide MySQL and PG examples.

**Comment by kuchaguangjie** [comment]
Probably it's worth mention that:

-  inside `dataSourceName`, `dbname=` must be put at end.
- Exactly one of `dataSourceName`'s `dbname=` and `dbName` should be provided, the other should be left empty.

Since, seems for pg, the code append `dbName`'s value to `dataSourceName`.

**e.g:**

This will work:
> dataSourceName = "user=postgres password=123456 sslmode=disable timezone=Asia/Shanghai dbname="
> dbName = casdoor

This will work too:
> dataSourceName = "user=postgres password=123456 sslmode=disable timezone=Asia/Shanghai dbname=casdoor"
> dbName =

This will get **error**:
> dataSourceName = "user=postgres password=123456 sslmode=disable dbname= timezone=Asia/Shanghai"

This will get **error** too:
> dataSourceName = "user=postgres password=123456 sslmode=disable timezone=Asia/Shanghai dbname=casdoor"
> dbName = casdoor

**Comment by hsluoyz** [comment]
@kuchaguangjie thanks for the comment!

**Pull request merged**